feat: disjunctive patterns

mzacho opened the pull request:
Extends the thir importer to handle or-patterns (`|`), see https://rust-lang.github.io/rfcs/2535-or-patterns.html.

The modification of the Thir in `engine/lib/ast.ml` is directly taken from `rustc_middle::thir`.

Running `cargo hax into coq` inside `tests/pattern-or` yields

```
Inductive t_E : Type :=
| E_At_E
| E_Bt_E.

Definition bar (x : t_E_t) : unit :=
  match x with
  | E_A  | E_B  => tt
  end.
```

If the types are adjusted (see #303) then I believe this is about what we want for the coq backend. See https://coq.inria.fr/refman/language/extensions/match.html#disjunctive-patterns.

Closes #161.

W95Psp reviewed (changes requested):
Looks good, thank you! :smiley: 

Some backends don't have support for nested disjunctive patterns (F*, but probable EC and others in the future), so I think we should feature-gate those (by adding a `DisjunctivePat` in `features.ml` and in the constructor `Or`).

So I'd suggest either:
 - add the feature;
 - or postpone that task, open an issue about it, and (for now) throw an error in the F* backend.

mzacho commented:
Hi @W95Psp - I realize your on vacation, please don't look at this until your back! :)

I attempted to hide disjunctive patterns behind a feature-gate as you proposed, however I ran into issues when trying to understand the current feature setup. Attempted to follow the structure of the setup to the best of my knowledge, however Ocaml reported errors which wasn't clear to me how to fix, so I gave up. 

These are the changes I made, before giving up:

```
modified   engine/backends/coq/coq/coq_backend.ml
@@ -38,7 +38,8 @@ module SubtypeToInputLanguage
              and type block = Features.Off.block
              and type for_loop = Features.Off.for_loop
              and type for_index_loop = Features.Off.for_index_loop
-             and type state_passing_loop = Features.Off.state_passing_loop) =
+             and type state_passing_loop = Features.Off.state_passing_loop
+             and type disjunctive_pat = Features.Off.disjunctive_pat) =
 struct
   module FB = InputLanguage
 
modified   engine/backends/easycrypt/easycrypt_backend.ml
@@ -63,6 +63,7 @@ module RejectNotEC (FA : Features.T) = struct
         let construct_base _ = Features.On.construct_base
         let for_loop = reject
         let for_index_loop _ = Features.On.for_index_loop
+        let disjunctive_pat = reject
 
         let metadata =
           Phase_utils.Metadata.make (Reject (NotInBackendLang EasyCrypt))
modified   engine/backends/fstar/fstar_backend.ml
@@ -37,7 +37,8 @@ module SubtypeToInputLanguage
              and type block = Features.Off.block
              and type for_loop = Features.Off.for_loop
              and type for_index_loop = Features.Off.for_index_loop
-             and type state_passing_loop = Features.Off.state_passing_loop) =
+             and type state_passing_loop = Features.Off.state_passing_loop
+             and type disjunctive_pat = Features.Off.disjunctive_pat) =
 struct
   module FB = InputLanguage
 
modified   engine/lib/ast.ml
@@ -281,7 +281,7 @@ functor
         }
       (* An or-pattern, e.g. `p | q`.
          Invariant: `List.length subpats >= 2`. *)
-      | POr of { subpats: pat list }
+      | POr of { witness: F.disjunctive_pat; subpats: pat list }
       | PArray of { args : pat list }
       | PDeref of { subpat : pat; witness : F.reference }
       | PConstant of { lit : literal }
modified   engine/lib/features.ml
@@ -21,7 +21,9 @@ loop,
   construct_base,
   monadic_action,
   monadic_binding,
-  block]
+  block,
+  disjunctive_pat
+]
 
 module Full = On
 
modified   engine/lib/import_thir.ml
@@ -731,7 +731,7 @@ end) : EXPR = struct
           in
           (c_constant_expr value |> pat_of_expr).p
       | Array _ -> unimplemented [ pat.span ] "Pat:Array"
-      | Or { pats } -> POr { subpats = List.map ~f:c_pat pats }
+      | Or { pats } -> POr { witness = W.disjunctive_pat; subpats = List.map ~f:c_pat pats }
       | Slice _ -> unimplemented [ pat.span ] "pat Slice"
       | Range _ -> unimplemented [ pat.span ] "pat Range"
     in
modified   engine/lib/print_rust.ml
@@ -184,7 +184,7 @@ module Raw = struct
           !"("
           & concat ~sep:!", " (List.map ~f:(fun { pat; _ } -> ppat pat) args)
           & !")"
-    | POr { subpats } -> concat ~sep:!" | " (List.map ~f:ppat subpats)
+    | POr { subpats; _ } -> concat ~sep:!" | " (List.map ~f:ppat subpats)
     | PArray { args } -> !"[" & concat ~sep:!"," (List.map ~f:ppat args) & !"]"
     | PDeref { subpat; _ } -> !"&" & ppat subpat
     | PConstant { lit } -> pliteral e.span lit
modified   engine/lib/subtype.ml
@@ -97,7 +97,7 @@ struct
             is_record;
             is_struct;
           }
-    | POr { subpats } -> POr { subpats = List.map ~f:dpat subpats }
+    | POr { subpats; witness } -> POr { subpats = List.map ~f:dpat subpats; witness = S.disjunctive_pat witness }
     | PArray { args } -> PArray { args = List.map ~f:dpat args }
     | PConstant { lit } -> PConstant { lit }
     | PBinding { mut; mode; var : Local_ident.t; typ; subpat } ->
```

Will gladly attempt to fix #306 if you can give me some pointers on how to proceed :)

W95Psp commented:
Hi,
I think we should merge that PR as it is, and have a follow-up PR if we want to feature gate things.
As discussed in #306, if we want to distinguish nested/non-nested disjunctive patterns, then adding a feature is probably overkill.

Regarding you changes, that seems just fine to me!
(OCaml gives really bad error messages when there is the slightest mistake in functor application though, so I suspect this is the kind of error you got)
